File: web/tests/detail-job.test.js
/**
 * Headless tests for web/js/lib/detail-job.js (WP 4a.4).
 *
 * Run: node --test "web/tests/*.test.js"   (see web/tests/README.md)
 */
import { test } from "node:test";
import assert from "node:assert/strict";
import { findTrackedJob, detailJobActions, DETAIL_JOB_ACTION } from "../js/lib/detail-job.js";

function job(overrides) {
  return { id: 1, appid: 440, type: "prefill", status: "running", ...overrides };
}

test("findTrackedJob: finds a queued/running/paused PREFILL job for the appid", () => {
  const jobs = [job({ status: "queued" })];
  assert.equal(findTrackedJob(jobs, 440), jobs[0]);
});

test("findTrackedJob: deliberately BROADER than lib/game-status.js's findLiveJob -- includes queued", () => {
  const jobs = [job({ id: 5, status: "queued" })];
  assert.equal(findTrackedJob(jobs, 440).id, 5);
});

test("findTrackedJob: excludes GC jobs -- pause/resume/download are prefill-only concepts", () => {
  const jobs = [job({ id: 9, type: "gc", status: "running" })];
  assert.equal(findTrackedJob(jobs, 440), undefined);
});

test("findTrackedJob: excludes a different appid, and a finished job", () => {
  assert.equal(findTrackedJob([job({ appid: 999 })], 440), undefined);
  assert.equal(findTrackedJob([job({ status: "done" })], 440), undefined);
});

test("findTrackedJob: handles a missing/non-array jobs snapshot", () => {
  assert.equal(findTrackedJob(undefined, 440), undefined);
  assert.equal(findTrackedJob(null, 440), undefined);
});

test("detailJobActions: mirrors api/README.md's job control table exactly", () => {
  assert.deepEqual(detailJobActions(job({ status: "queued" })), new Set([DETAIL_JOB_ACTION.CANCEL]));
  assert.deepEqual(
    detailJobActions(job({ status: "running" })),
    new Set([DETAIL_JOB_ACTION.PAUSE, DETAIL_JOB_ACTION.CANCEL]),
  );
  assert.deepEqual(
    detailJobActions(job({ status: "paused" })),
    new Set([DETAIL_JOB_ACTION.RESUME, DETAIL_JOB_ACTION.CANCEL]),
  );
});

test("detailJobActions: no actions for a finished job or no job at all", () => {
  assert.deepEqual(detailJobActions(job({ status: "done" })), new Set());
  assert.deepEqual(detailJobActions(job({ status: "error" })), new Set());
  assert.deepEqual(detailJobActions(job({ status: "cancelled" })), new Set());
  assert.deepEqual(detailJobActions(null), new Set());
  assert.deepEqual(detailJobActions(undefined), new Set());
});

File: app/app/src/test/java/dev/steamvault/app/ui/detail/logic/DetailJobTest.kt
package dev.steamvault.app.ui.detail.logic

import dev.steamvault.app.net.model.JobSummary
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test

class DetailJobTest {

    private fun job(id: Int, appid: Int, type: String, status: String) = JobSummary(
        id = id, appid = appid, type = type, status = status, created_at = "2026-08-01T00:00:00Z",
    )

    @Test
    fun `finds a queued prefill job for the app -- broader than the grid card's findLiveJob`() {
        val jobs = listOf(job(1, 440, "prefill", "queued"))
        assertEquals(jobs[0], findTrackedJob(jobs, 440))
    }

    @Test
    fun `finds a running and a paused prefill job`() {
        assertEquals("running", findTrackedJob(listOf(job(1, 440, "prefill", "running")), 440)?.status)
        assertEquals("paused", findTrackedJob(listOf(job(1, 440, "prefill", "paused")), 440)?.status)
    }

    @Test
    fun `ignores a GC job for the same app -- GC jobs are not job-control targets here`() {
        val jobs = listOf(job(1, 440, "gc", "running"))
        assertNull(findTrackedJob(jobs, 440))
    }

    @Test
    fun `ignores a finished job and a job for a different app`() {
        assertNull(findTrackedJob(listOf(job(1, 440, "prefill", "done")), 440))
        assertNull(findTrackedJob(listOf(job(1, 730, "prefill", "running")), 440))
    }

    @Test
    fun `job control table -- queued offers cancel only`() {
        assertEquals(setOf(DetailJobAction.CANCEL), detailJobActions(job(1, 440, "prefill", "queued")))
    }

    @Test
    fun `job control table -- running offers pause and cancel`() {
        assertEquals(setOf(DetailJobAction.PAUSE, DetailJobAction.CANCEL), detailJobActions(job(1, 440, "prefill", "running")))
    }

    @Test
    fun `job control table -- paused offers resume and cancel`() {
        assertEquals(setOf(DetailJobAction.RESUME, DetailJobAction.CANCEL), detailJobActions(job(1, 440, "prefill", "paused")))
    }

    @Test
    fun `job control table -- no job, or a finished job, offers nothing`() {
        assertEquals(emptySet<DetailJobAction>(), detailJobActions(null))
        assertEquals(emptySet<DetailJobAction>(), detailJobActions(job(1, 440, "prefill", "done")))
    }
}
